lib/core/src/text.rs

```diff
@@ -1,65 +1,150 @@
+
+/// Text component builder
+pub struct TextComponentBuilder<'a, Str:AsRef<str>> {
+    pub text     : Str,
     pub font     : &'a mut FontRenderInfo,
     pub position : Vector2<f64>,
     pub size     : f64,
     pub color    : Color<f32>,
 }
 
-#[derive(Debug)]
-pub struct TextComponent {
-    gl_context                    : WebGlRenderingContext,
-    gl_program                    : Program,
-    gl_vertex_buffer              : WebGlBuffer,
-    gl_texture_coordinates_buffer : WebGlBuffer,
-    gl_msdf_texture               : WebGlTexture,
-    buffers_size                  : usize,
-}
+impl<'a,Str:AsRef<str>> TextComponentBuilder<'a,Str> {
 
-impl<'a> TextComponentBuilder<'a> {
-    pub fn build(mut self, workspace  : &Workspace) -> TextComponent {
+    /// Build a new text component rendering on given workspace
+    pub fn build(mut self, workspace : &Workspace) -> TextComponent {
         self.load_all_chars();
-        let gl_context           = workspace.context.clone();
-        let gl_program           = self.create_program(&gl_context);
-        let gl_vertex_buffer     = self.create_vertex_bufffer(&gl_context);
-        let gl_tex_coord_buffer  = self.create_texture_coordinates_buffer(&gl_context);
-        let gl_msdf_texture      = self.create_msdf_texture(&gl_context);
-        let glyph_vertices_count = glyph_render::GLYPH_SQUARE_VERTICES_BASE_LAYOUT.len();
-        let buffers_size         = self.text.len() * glyph_vertices_count;
+        let displayed_lines   = 100; // TODO[AO] replace with actual component size
+        let displayed_columns = 100;
+        let gl_context        = workspace.context.clone();
+        let gl_program        = self.create_program(&gl_context);
+        let gl_msdf_texture   = self.create_msdf_texture(&gl_context);
+        let lines             = self.split_lines();
+        let mut buffers       = TextComponentBuffers::new(&gl_context,displayed_lines,displayed_columns);
+        let to_window         = self.to_window_transform();
+
         self.setup_uniforms(&gl_context, &gl_program);
+        buffers.refresh_all(&gl_context, &lines, self.font, &to_window);
         TextComponent {
             gl_context,
             gl_program,
-            gl_vertex_buffer,
-            gl_texture_coordinates_buffer : gl_tex_coord_buffer,
             gl_msdf_texture,
-            buffers_size
+            buffers,
+            lines,
+            to_window,
         }
     }
 
     fn load_all_chars(&mut self) {
-        for ch in self.text.chars() {
+        for ch in self.text.as_ref().chars() {
             self.font.get_glyph_info(ch);
         }
     }
 
+    fn split_lines(&self) -> Vec<Line> {
+        let lines_text = self.text.as_ref().split('\n');
+        let lines_iter = lines_text.map(Self::initialize_line);
+        lines_iter.collect()
+    }
+
+    fn initialize_line(text:&str) -> Line {
+        Line {
+            content : text.to_string(),
+        }
+    }
```

Comment:
Very minor thing, but I think we should add it to the style guide:
```rust
    fn initialize_line(text:&str) -> Line {
        let content = text.to_string();
        Line {content}
    }
```

Reply:
Well, this is the one I don't agree with. I don't see the point of creating let here, and in more complex case it is nice to see clearly what is assigned to a field (without quicksearchng for label name).

I see the point in doing this for function arguments, where reader does not see what the arguments are without jumping to definition. But here the field names are mentioned, so this is not the case.

In other words, making labels for fields arguments is like doing this:
```rust
fn abs(x:usize) -> usize {
     let result = if x >= 0 {x} else {-x};
    result
}
```

Reply:
I'd like to make this rule: don't male labels for structure fields if it is constructed as the return of the function (if you construct structure in the middle of function body you probably need to split this function anyway)